internal/testutils/networking/l2_segment_linux.go

```diff
@@ -0,0 +1,391 @@
+type L2Segment struct {
+	BridgeNs    Netns
+	bridgeNlh   *netlink.Handle
+	bridgeIface netlink.Bridge
+	nextIP      netip.Addr
+	mask        net.IPMask
+}
+
+func prefixMask(p netip.Prefix) net.IPMask {
+	return net.CIDRMask(p.Bits(), p.Addr().BitLen())
+}
+
+func NewL2Segment(t *testing.T, brName string, subnet netip.Prefix, testID uint32, reusePrevious bool) (*L2Segment, error) {
+	t.Helper()
+
+	sgmt := &L2Segment{
+		nextIP: subnet.Addr().Next(),
+		mask:   prefixMask(subnet),
+	}
+
+	var err error
+	sgmt.BridgeNs = NewNamedNetns(t, fmt.Sprintf("bridge-%8x", testID), reusePrevious)
+	if err != nil {
+		return sgmt, fmt.Errorf("could not create bridge netns: %w", err)
+	}
+
+	if sgmt.bridgeNlh, err = sgmt.BridgeNs.NetlinkHandle(); err != nil {
+		return sgmt, fmt.Errorf("could not get netlink handle for netns %s: %w", sgmt.BridgeNs, err)
+	}
+
+	sgmt.bridgeIface = netlink.Bridge{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:  brName,
+			Flags: net.FlagUp,
+		},
+	}
+
+	// If reusePrevious is true, we don't need to create a new bridge interface if it already exists
+	_, err = sgmt.bridgeNlh.LinkByName(brName)
+	if err != nil {
+		var linkNotFoundErr netlink.LinkNotFoundError
+		if !errors.As(err, &linkNotFoundErr) {
+			assert.NilError(t, err)
+		}
+		if err := sgmt.bridgeNlh.LinkAdd(&sgmt.bridgeIface); err != nil {
+			return sgmt, fmt.Errorf("failed to create bridge interface: %w", err)
+		}
+	}
+	if err == nil && !reusePrevious {
+		t.Logf("Bridge %s already exists whereas TEST_REUSE_L2SEGMENT is empty.", sgmt.bridgeIface.Name)
+		t.FailNow()
+	}
+
+	return sgmt, nil
+}
+
+func (sgmt *L2Segment) AddHost(host *L3Host, reusePrevious bool) error {
+	host.IPAddr = sgmt.nextIP
+	host.IPMask = sgmt.mask
+	sgmt.nextIP = host.IPAddr.Next()
+
+	err := host.createVeth(sgmt.bridgeNlh, sgmt.bridgeIface, reusePrevious)
+	if err != nil {
+		return fmt.Errorf("failed to create veth: %w", err)
+	}
+
+	return nil
+}
+
+func (sgmt *L2Segment) Destroy(logger testutils.Logger) {
+	sgmt.BridgeNs.Destroy(logger)
+}
+
+// L3Host represents a host in a L2Segment. The host might be a dedicated net namespace or the net namespace where
+// the test is executed.
+type L3Host struct {
+	// Ns is the Netns representing this host.
+	Ns Netns
+	// HostIfaceName is the name of the veth interface that resides on the host Ns.
+	HostIfaceName string
+	// BridgedIfaceName is the name of the veth interface associated to HostIfaceName. It might not reside in the host Ns.
+	BridgedIfaceName string
+	MACAddr          net.HardwareAddr
+	IPAddr           netip.Addr
+	IPMask           net.IPMask
+}
+
+// createVeth creates a veth pair. The interface being named after hostIfaceName is put into hostNs, and the one being
+// named after bridgedIfaceName in BridgeNs. The "host" side receives the addr IP address. It returns the MAC address
+// of the "host" side, a function to remove this interface and an error, if any. The veth pair should be ready to use
```

Comment:
This comment is totally out of touch with what the code does now. I totally forgot it was on my todo-list of things to fix before moving the PR out of draft state.